### DISCUSSION TITLE
Smoother magit-commit-fixup and magit-commit-instant-fixup

### DISCUSSION THREAD

**Discussion by ghost** [magit/magit#4915] [Ideas]
I do a lot of grooming of  commit series using magit-commit-instant-fixup.  Each time I issue 'cF' magit opens a window showing the commits on my branch.  Many of those commits have already been push upstream.  The expectation is that I will move point to the target commit and hit ^C ^C.

Unless I consciously want to rewrite upstream history, all of my fixups are going to be applied to the commits in my "unmerged into <upstream>" section.  If I have that section expanded and I position point on my intended target commit, then I observe that 'cF' lands me on that exact commit in the popped up branch window.  I still have to experience the popping up and deletion of that window and the concomitant shift in focus.

I suggest a  new commit variant.  Let's call it unmerge-fixup.  This would be available only if the "unmerged into <upstream>" section is present.  Its behavior would be something like:

* Remember the expanded/collapsed state of the "unmerged into <upstream>" section
* Expand the section if needed
* Move to the most recent target commit (or most recent if that commit is not present)
* Wait for user's ^C ^C on a desired target
* (Error if point is not on a commit in the "unmerged into <upstream>" section.)
* Remember this most recent commit
* Carry out the fixup
* If needed, collapse the "unmerged into <upstream>" section

**Comment by tarsius** [comment]
The non-instant variants by default do not require confirmation. Or in other words the don't pop up the diff and log. (Customize `magit-commit-squash-confirm` to *disable* that.) The instant variants don't support that because it would be dangerous. If you squash to wrong changes and/or squash them into the wrong commit, then it is unlikely you will notice until after you have made other changes. If you notice later, it will be very hard to correct the mistake.

So the mentioned "shift in focus" is actually intentional, it gives you a change to notice a mistake.

I encourage you to implement a variant along the lines of what you are proposing yourself, but I don't think I will add something like that to Magit itself.

Also, are you aware of `magit-commit-autofixup`? It is available from `magit-popup` but hidden by default. That command can often automatically determine the commit to which the changes should be applied.
